### PR TITLE
feat(version-file): add version file support

### LINKED ISSUE #29: Transition application versioning from versioninfo to Zephyr's VERSION file
## Context

Currently many of our projects use [irnas-versioninfo-lib](https://github.com/IRNAS/irnas-versioninfo-lib). This library has served us well, as it was automatically running `git describe` command whenever the build ran. The library then parsed the output of the `git describe` , extracted specific parts and converted them into the C macros.

The library was developed at the time when Zephyr had now specific application versioning system. In the past year that changed as Zephyr added support for the [VERSION](https://docs.zephyrproject.org/latest/build/version/index.html) file.

Since it seems that all version specific logic in the Zephyr and MCUboot will look at the macros that are derived from this `VERSION` file, it is in our interest to ditch the `versioninfo` and start using the `VERSION` file.

However, one drawback that the VERSION file has is the fact that there is no update mechanism. Application developers need to update the contents either manually or write a script that does it for them.

@TjazVracko already wrote such a short python script that does that. This is issue was created so we can think about all possible usecases and how to integrate this file in the new/old projects in the most seamless way.  

## Required steps / Implementation details

* The update scrip can be a part of the Zephyr template, it should be located in the `scripts/` folder.
* It should take an path argument to the version file that needs to be updated. As discussed in the IRL with @TjazVracko there is probably no need that it also needs an option to take the version as an another argument. Parsing the `git describe` should be enough.
* We need to think about what is the expected usecase within the CI.
* We need to think about what is the expected usecase when developing locally.
* We need a working demo setup in this repository and whatever instructions that are needed for developers to use this script.
* Decide what the fields in the VERSION mean to the IRNAS team and document that.

I would be very happy, if the CI workflow files **do not need to change** and that script only needs to be called by a specific make target. Then this is in my mind the easiest change to the existing infra that we have setup.

## Definition of Done

A new working versioning workflow with the VERSION file and documentation describing relevant details.

### PR DESCRIPTION
**Pull request by TjazVracko**
# Description

This commit removes the use of the version-info lib and replaces it
with a python script and Zephyr's VERSION file system.

The python script version.py can be used to generate a valid VERSION
file based on the provided tag or based on the output of git describe.

The app firmware no linger uses the version-info lib and instead reads
the version from app_version.h, ncs_version.h, and other configs.

Closes #29

## Areas of interest for the reviewer

All.

I have added the script to generate VERSION file and updated the main app.
We still require updates to the CI workflows as discussed in #29.
@MarkoSagadin, I am handing it off to you for the CI updates.

I am also unsure if the .gitigore and the makefile are correct. That is based on how the CI will handle the VERSION files.
Keep in mind that if there are 2 apps, 2 version files have to be generated, one at the root of each "zephyr project".

Another note: The script now silently fails and sets the version to "0.0.0-0" (as described by the help text). I am not sure if we want this. Maybe it should fail if the tag is invalid. Thoughts?

## Checklist

- [x] My code follows the [style guidelines] as defined by IRNAS.
- [x] I have performed a self-review of my code.
- [x] My changes generate no new warnings.
- [x] I added/updated source code documentation for all newly added or changed functions.
- [x] ~~I updated all customer-facing technical documentation.~~ - This PR
      introduced only internal facing changes.

## After-review steps

- Reviewer can merge and delete the branch.

[style guidelines]: https://github.com/IRNAS/irnas-guidelines-docs/blob/main/docs/developer_guidelines.md

### CODE REVIEW

File: scripts/version/version.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        required=True,
+        help="The file path to write the version information to.",
+    )
+    parser.add_argument(
+        "-t",
+        "--tag",
+        metavar="TAG",
+        type=str,
+        required=False,
+        help='The tag to use for version information in the format "vMAJOR.MINOR.PATCH-TWEAK". If not provided, git decsribe will be used.',
```
Comment (MarkoSagadin) [inline scripts/version/version.py:64]: Nit: I would add the angled braces and example input to better present the required format (that leading "v" is a bit hidden).

```suggestion
        help='The tag to use for version information in the format "v<MAJOR>.<MINOR>.<PATCH>-<TWEAK>", eg. v1.2.3.-4. If not provided, git describe will be used.',
```

**Review by MarkoSagadin — CHANGES_REQUESTED** [review]
Minor remarks and nits.

**Comment by MarkoSagadin** [comment]
@TjazVracko 

I went through the PR and I have some thoughts. I know that we had some discussion IRL as well as in issue #29, but it think that not everything is written down.

> Keep in mind that if there are 2 apps, 2 version files have to be generated, one at the root of each "zephyr project".

That is ok, but please mention this next to the `gen-version` target, see codechecker targets for inspiration.

> I am also unsure if the .gitigore and the makefile are correct. That is based on how the CI will handle the VERSION files.

I know that I was talking in the linked issue about adding an environment variable for the version (and then ignored you when you wanted to know the name, sorry for that :sweat_smile: ), but seeing this I have a slightly new idea now.

What if:
* VERSION file is ignored (like it is now)
* Template comes with the version file set to some value, could be "0.0.0-0" or "99.99.99-99" or something else, just to indicate the development state. (This means that we don't have to run `make gen-versoin` at startup, something that I dislike)
* In the CI script doesn't need fetch any value from the env var, it just runs `git describe` as a part of the `make pre-build` command.

I think that with this setup:
- We get a distinct version number for the development
- No extra command at project setup
- No requred change to the existing CI scripts
- VERSION file in the project is always the same (the default value) but can be overwritten for the dev purposes

Does this makes sense? Am I missing some corner case?

> Another note: The script now silently fails and sets the version to "0.0.0-0" (as described by the help text). I am not sure if we want this. Maybe it should fail if the tag is invalid. Thoughts?

So, I thought about this. The script should probably fail, if the given tag (or tag parsed from the `git describe`) is not in the correct format. If I am running this in the CI I want to know about the possible configuration mistake as soon as possible, no when then devices are in the production and I am somewhat depending on the version numbers.

**Review by TjazVracko — COMMENTED** [review]
@MarkoSagadin I changed some things based on your suggestions:

1. added a default VERSION file
2. the script now errors out of the tag is invalid (bad command line argument or no git tag)
3. The makefile has a bit more of an explanation

My only concern is that, when someone uses the template to create a project with 2 apps, he will copy and paste the app dir, and when committing it, will not "force add" the ignored VERSION file.
I would still like it if the version file was updated and committed by CI.

**Comment by MarkoSagadin** [comment]
> @MarkoSagadin I changed some things based on your suggestions:

Thank you!

> My only concern is that, when someone uses the template to create a project with 2 apps, he will copy and paste the app dir, and when committing it, will not "force add" the ignored VERSION file.

That also crossed my mind, it is a reasonable concern. I would mitigate this by adding a entry to the main checklist, that link to the scripts/version/README.md which explains the use of VERSION file and the multiple app setup. 

> I would still like it if the version file was updated and committed by CI.

Can you explain why? I think that if this is done, then what we gain is only the removal of the VERSION file from the `.gitignore`, which would then avoid the above mentioned "force add" situation, is that right?

I am against this change as currently the release part of the CI is really only concerned with updating the CHANGELOG entry, which is completely project agnostic, and all makefile related targets are used in other workflows.  I would like to stay it that way, it was designed with this idea in mind.

---

There is also another possible situation:
* VERSION file is still in `.gitignore`
* It is not provided in the template.

`make gen-version` still works in its current form, as well as the CI. But I would guess that MCUboot might complain, or there might be some cases during the development where this might not be desirable? 

What do you think?

**Review by MarkoSagadin — COMMENTED** [review]
Done changes are ok, but we still need to figure out this.

**Comment by TjazVracko** [comment]
> I am against this change as currently the release part of the CI is really only c...

There are 2 problems with the committed VERSION file that is also ignored:

FIRST:
1. VERSION file is ignored and added to the template
2. I am developing something that needs VERSION changes, so I run the script locally to change the file.
3. running `git status` now shows the VERSION file as changed
4. running `git add app/` stages the VERSION file - I might accidentally commit it.

This is annoying...., but probably won't come up often.

SECOND:
this also means that the git state will be dirty when CI runs the script before generating artefacts with `east release`. This is not ideal, since they will be named wrong.

---------------

This means that not adding a VERSION file and ignoring it is better. But the problem now is that the build fails if no version file exists (`app_version.h` is not generated). This can be fixed by running `make gen-version` locally once.

So I think the "correct" solution is actually:
1. add VERSION to .gitignore
2. do not commit it to the template
3. Modify the version script so it can generate a "dev" VERSION file, even if no git tag exists (since currently make `gen-version` fails if there is no release tag, which happens when a project is new)
4. have a make command that runs the dev option described above
5. add the call to that make command in the repo setup instructions.


We should discuss this in person a bit more, It requires a demonstration.

**Comment by TjazVracko** [comment]
@MarkoSagadin You can also modify common.conf and main.c (remove most things) and build for native_posix so you can try it out yourself.

**Comment by TjazVracko** [comment]
@MarkoSagadin 

Git seeing the changes to the VERISON file can be avoided using `git update-index --assume-unchanged`, but this would have to be run in CI and on every developer machine, so this is also not ideal.

File: app/CMakeLists.txt [hunk 2]
```diff
@@ -2,6 +2,8 @@ cmake_minimum_required(VERSION 3.13.1)
 
 find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
 
+zephyr_compile_definitions(CONFIG_BUILD_TYPE="${EAST_BUILD_TYPE}")
```
Comment (TjazVracko) [inline app/CMakeLists.txt:5]: @MarkoSagadin Please also think about if we are ok injecting the east build type like this.
I added this here to be able to print all of the same information that `version-info` did. But it is not very important. 

Any thoughts?

Comment (MarkoSagadin) [reply app/CMakeLists.txt:5]: I would remove this.

**Pull request merged**